File: StoreHours_Java/StoreHours.java
/**
 * Copyright 2024 David Lira
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import java.time.DayOfWeek;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.util.Map;

public class StoreHours {
    private final Map<String, String[]> hours;
    private final Map<String, String[]> exceptions;

    public StoreHours(Map<String, String[]> hours, Map<String, String[]> exceptions) {
        this.hours = hours;
        this.exceptions = exceptions;
    }

    public String[] hours() {
        DayOfWeek currentDayOfWeek = DayOfWeek.from(java.time.LocalDate.now());
        String currentDayKey = currentDayOfWeek.toString().substring(0, 3).toLowerCase();

        if (hours.containsKey(currentDayKey)) {
            return hours.get(currentDayKey);
        } else {
            return new String[]{};
        }
    }

    public String[] exceptions() {
        java.time.ZoneId madridZone = java.time.ZoneId.of("Europe/Madrid");
        java.time.LocalDate currentDate = java.time.LocalDate.now(madridZone);
        String formattedDate = currentDate.format(DateTimeFormatter.ofPattern("M/d"));

        if (exceptions.containsKey(formattedDate)) {
            return exceptions.get(formattedDate);
        } else {
            return new String[]{};
        }
    }

    public boolean isOpen() {
        String[] regularHours = hours();
        String[] exceptions = exceptions();

        if (regularHours != null && regularHours.length > 0 && !regularHours[0].isEmpty()) {
            java.time.LocalTime currentTime = java.time.LocalTime.now();
            java.time.format.DateTimeFormatter timeFormatter = java.time.format.DateTimeFormatter.ofPattern("HH:mm");

            for (String exception : exceptions) {
                String[] times = exception.split("-");
                String start = times[0].trim();
                String end = times[1].trim();

                if (isWithinTimeRange(currentTime, LocalTime.parse(start, timeFormatter), LocalTime.parse(end, timeFormatter))) {
                    return false;
                }
            }

            for (String hours : regularHours) {
                String[] times = hours.split("-");
                String start = times[0].trim();
                String end = times[1].trim();

                if (isWithinTimeRange(currentTime, LocalTime.parse(start, timeFormatter), LocalTime.parse(end, timeFormatter))) {
                    return true;
                }
            }
        }

        return false;
    }

    private boolean isWithinTimeRange(LocalTime currentTime, LocalTime start, LocalTime end) {
        return !currentTime.isBefore(start) && !currentTime.isAfter(end);
    }
}

File: StoreHours_PHP/storehours.php
<?php

class StoreHours {
  private $hours;
  private $exceptions;

  public function __construct($hours, $exceptions) {
      $this->hours = $hours;
      $this->exceptions = $exceptions;
  }

  public function hours() {
      $currentDayOfWeek = date('N');
      $currentDayKey = strtolower(date('D'));

      if (isset($this->hours[$currentDayKey])) {
          return $this->hours[$currentDayKey];
      } else {
          return [];
      }
  }

  public function exceptions() {
    date_default_timezone_set('Europe/Madrid');
    $currentDate = date('n/j');

    if (isset($this->exceptions[$currentDate])) {
        return $this->exceptions[$currentDate];
    } else {
        return [];
    }
}

  public function isOpen() {
    date_default_timezone_set('Europe/Madrid');

    $regularHours = $this->hours();
    $exceptions = $this->exceptions();

    if (!is_null($regularHours) && is_array($regularHours) && !empty($regularHours) && !empty($regularHours[0])) {
       $currentTime = date('H:i');
      
      if (!empty($exceptions)) {
          foreach ($exceptions as $exception) {
              list($start, $end) = explode('-', $exception);
              if ($this->isWithinTimeRange($currentTime, $start, $end)) {
                  return false;
              }
          }
      }

      foreach ($regularHours as $hours) {
          list($start, $end) = explode('-', $hours);
          if ($this->isWithinTimeRange($currentTime, $start, $end)) {
              return true;
          }
      } 
    }

    return false;   
  }

  private function isWithinTimeRange($time, $start, $end) {
    $currentTime = date('H:i', strtotime($time));
    $startTime = date('H:i', strtotime($start));
    $endTime = date('H:i', strtotime($end));

    return ($currentTime >= $startTime && $currentTime <= $endTime);
  }

  public function __toString() {
    return 'StoreHours Object';
  }
}
?>